### PR TITLE
clang: update to 14 and bionic

### PR DESCRIPTION
**Pull request by kyessenov**

### CODE REVIEW

**Comment by ericvn** [comment]
Are you planning to do any more investigation before we merge this, or with the resulting image before we move to it in the pipeline?

File: docker/build-tools/Dockerfile [hunk 1]
```diff
@@ -661,28 +661,29 @@ ENTRYPOINT ["/usr/local/bin/docker-entrypoint"]
 # Clang+LLVM
 ##############
 
-FROM ubuntu:xenial AS clang_context_amd64
-FROM ubuntu:bionic AS clang_context_arm64
+FROM ubuntu:20.04 AS clang_context_amd64
+FROM ubuntu:20.04 AS clang_context_arm64
```
Comment (howardjohn) [inline docker/build-tools/Dockerfile:665]: nit: we only needed 2 clauses here when they were different

**Review by howardjohn — COMMENTED** [review]
Per https://github.com/istio/tools/pull/2246, updating the OS here ends up being a breaking change to VM users due to glibc.

In https://github.com/istio/tools/pull/2246#issuecomment-1315712870 I listed some possible ways to lessen this which I don't think we have done.

I think we also have the choice of 18 vs 20? Or does 18 cause problems? I know 18 just went EOL this month but some folks have extended support, etc.

I do think we should do this just need to be careful with how we roll it out

**Comment by kyessenov** [comment]
I don't think we are going to have a choice not to update soon. C++20 is coming and we need newer compilers to prevent build errors.

**Comment by kyessenov** [comment]
@howardjohn Are you sure that glibc is linked into Envoy? clang needs newer base but that doesn't mean the resulting binary needs glibc.

**Comment by howardjohn** [comment]
Yeah I agree but we also need to acknowledge this is a breaking change not
just an internal change to our build environment and make sure that is
conveyed to customers + mitigated appropriately.

**Comment by howardjohn** [comment]
> @howardjohn Are you sure that glibc is linked into Envoy? clang needs newer base but that doesn't mean the resulting binary needs glibc.

As far as I know yes, that is why we use ubuntu 16 currently. And why we build the "centos 7 compatible envoy" on centos 7.

**Comment by kyessenov** [comment]
I am not comfortable releasing something that's not well tested. Did we actually run Envoy tests on older platforms? Upstream CI is only testing with clang 14 and focal.

**Comment by kyessenov** [comment]
I wonder if the issue is just libc++ using recent GLIBC ABI.

**Comment by howardjohn** [comment]
We do run tests on centos7, debian11, ubuntu xenail, and ubuntu jammy. Not
100% of the tests or anything but we do have some minimal coverage

**Comment by kyessenov** [comment]
Not Istio tests - Envoy tests. Upstream doesn't test on centos at all - so we don't know if they pass. There are many cases where small variations in the compiler/runtime causes bugs (on windows and mac, for example).

**Comment by kyessenov** [comment]
Downgraded amd64 to bionic.

**Comment by jacob-delgado** [comment]
We can also upgrade baselik too.

**Comment by jacob-delgado** [comment]
There's no rush. I just saw this as an opportunity to do this as well. I'm not too familiar with all scenarios tested. We don't have any users running on VMs on old versions of Centos 7 so upgrading LLVM would be a great addition if possible.

**Comment by jacob-delgado** [comment]
> Per #2246, updating the OS here ends up being a breaking change to VM users due to glibc.
> 
> In [#2246 (comment)](https://github.com/istio/tools/pull/2246#issuecomment-1315712870) I listed some possible ways to lessen this which I don't think we have done.
> 
> I think we also have the choice of 18 vs 20? Or does 18 cause problems? I know 18 just went EOL this month but some folks have extended support, etc.
> 
> I do think we should do this just need to be careful with how we roll it out

At what point do we need to support "End of Standard Support" releases for Ubuntu? Bionic now requires an extended support contract as does Xenial. While users can still download them, they shouldn't be using them in production anymore (we know there will always be people lagging behind).

I'm of the opinion that the open source should try it's best to support versions still under extended support contracts, but is under no obligation to if it blocks progress. When that happens users should either use older versions of Istio (not suggested) or go with a vendor that will support it past its due date. I'm in fan of us updating to focal since that's what upstream Envoy is using in our build images myself. Given that would be 2 version bumps going to bionic is more than fine for now until a later time.

Thoughts @howardjohn and @ericvn? Should we at least message some of this to our users?

**Comment by howardjohn** [comment]
In the WG call I think we had agreed to send a deprecation notice in one
release, bump it in the next, and add documentation and tests for running
in docker.

I tried to put some effort on the later but it was a bit trickier than I
thought so I didn't actually complete it. I think documenting it is easier
than testing it in our normal CI -- but the istio.io docs tests may
actually be easier to write

**Comment by shane-yuan** [comment]
ubuntu:bionic doesn't have libncurses6, which is needed to build the https://github.com/envoyproxy/envoy/pull/27087.

```
%  docker run -it --entrypoint "/bin/bash" ubuntu:bionic
root@9db05333f5f6:/# apt-get install libncurses6
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libncurses6

% docker run -it --entrypoint "/bin/bash" ubuntu:19.04
root@269ccc8c7e41:/# apt-get install libncurses6
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libncurses6 is already the newest version (6.1+20181013-2ubuntu2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

When I tried to build https://github.com/istio/istio/commit/2f62d77ad7359ca656fdde41afc3e5e3913662b0 by adding `build --define boringssl=fips` to .bazelrc and running build `BUILD_WITH_CONTAINER=1 make build`, I got the following error.
```
clang: error while loading shared libraries: libtinfo.so.6: cannot open shared object file: No such file or directory
```
The same build completed successfully on a Ubuntu v22 VM. Ubuntu has libncurses6 support since v19. I think we need to upgrade clang_context_amd64 to ubuntu:eoan to support building FIPS envoy in builder container.

**Comment by howardjohn** [comment]
@shane-yuan I am not an expert here but I am pretty sure FIPS is commonly used with the current build-tools, so I suspect you may be doing something different... not sure though. In any event, we will not update to eoan, only to LTS releases as this PR does.

**Comment by shane-yuan** [comment]
@howardjohn I'm able to build 1.180 GA with https://github.com/istio/istio/commit/697f4ff0fe6ee531bbd4f5f2a6b4b1f302c955a8 using the exactly the same setup. The build issue started after istio picked up https://github.com/envoyproxy/envoy/pull/27087 with a new version of FIPS boringssl.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**